log-manager/src/main/java/io/airlift/log/SocketHandler.java

```diff
@@ -0,0 +1,79 @@
+package io.airlift.log;
+
+import org.weakref.jmx.Managed;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.logging.ErrorManager.GENERIC_FAILURE;
+
+public class SocketHandler
+        extends BufferedHandler
+{
+    private static final int CONNECTION_TIMEOUT_MILLIS = 100;
+
+    private final InetSocketAddress socketAddress;
+    private final AtomicLong failedConnections = new AtomicLong(0);
+    @GuardedBy("this")
+    private Socket socket;
+
+    private SocketHandler(String logPath, Format format)
+    {
+        super();
+        try {
+            URI uri = new URI(logPath);
+            this.socketAddress = new InetSocketAddress(uri.getHost(), uri.getPort());
+        }
+        catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+        setFormatter(format.getFormatter());
+    }
+
+    public static SocketHandler createSocketHandler(String logPath, Format format)
+    {
+        SocketHandler handler = new SocketHandler(logPath, format);
+        handler.start();
+        return handler;
+    }
+
+    @Override
+    protected void preWriteCheck(byte[] message)
+    {
+        if (socket == null || socket.isClosed()) {
+            try {
+                socket = new Socket();
+                socket.connect(socketAddress, CONNECTION_TIMEOUT_MILLIS);
+                this.replaceCurrentOutputStream(socket.getOutputStream());
+            }
+            catch (IOException e) {
+                failedConnections.incrementAndGet();
+                reportError("Error connecting via socket to " + socketAddress.getHostName() + " on port " + socketAddress.getPort(), e, GENERIC_FAILURE);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        super.close();
+        try {
+            socket.close();
```

Comment:
It is not safe to close here.  The thread may still be writing messages.  This is why the old code closed in the thread cleanup loop.  If you switch to the interface I suggested, it will work the old way.